---
title: "Calculating JCC scores"
author: 
- name: Charlotte Soneson
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{Calculating JCC scores}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: jcc.bib
output: BiocStyle::html_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

This vignette describes how to use the `jcc` package to calculate JCC scores [@soneson2018jcc] for a given set of genes. 
The following files are necessary to run the entire workflow, and will be used as described below:

- a BAM file with reads aligned to a reference genome. 
The package has been tested with BAM files generated by the STAR aligner [@dobin2013star].
- a `data.frame` with the number of uniquely mapping and multimapping reads aligned across each annotated splice junction. 
Below we show how to generate such a data frame from the `SJ.out.tab` text file generated by STAR. 
- A `BSgenome` object for the genome that the reads in the BAM file were aligned to.
- A `gtf` file corresponding to the `BSgenome` and the BAM file.
- A transcript-to-gene conversion `data.frame` with at least two columns, named `tx` and `gene`. 
Additional columns providing, e.g., gene symbols can be included.
- A `data.frame` with estimated transcript abundances, with three columns named `transcript`, `TPM` and `count`. 
Note that this table should _not_ contain a column named `gene`.

The general procedure is outlined in the figure below. Please refer to the paper [@soneson2018jcc] for a more detailed description and motivation of the different steps. 

<img src="study_design.png" alt="drawing" width="400px"/>

In the rest of this vignette, we will illustrate each step using a small example data set provided with the package. 

# Fit fragment bias model

The first step in the process is to fit a fragment bias model, using the `r Biocpkg("alpine")` package. 
The `fitAlpineBiasModel()` function provides a wrapper containing the necessary steps. 
Note that the parameters of this function need to be adapted to each specific data set. 
See the help file and the `r Biocpkg("alpine")` package for more detailed explanations of each parameter. 
The `BSgenome` object (`Hsapiens`) should be replaced with the appropriate object for your organism and alignment.

```{r}
suppressPackageStartupMessages({
  library(jcc)
  library(dplyr)
  library(BSgenome.Hsapiens.NCBI.GRCh38)
})

gtf <- system.file("extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz", 
                   package = "jcc")
bam <- system.file("extdata/reads.chr22.bam", package = "jcc")

biasMod <- fitAlpineBiasModel(gtf = gtf, bam = bam, organism = "Homo_sapiens",
                              genome = Hsapiens, genomeVersion = "GRCh38",
                              version = 90, minLength = 230, maxLength = 7000,
                              minCount = 10, maxCount = 10000, subsample = TRUE,
                              nbrSubsample = 30, minSize = NULL, maxSize = 220, 
                              verbose = TRUE)
names(biasMod)
```

# Predict transcript coverage profiles

Next, we predict the coverage profile for each transcript in the specified genes. 
The `genes` argument can be set to `NULL` to predict the coverage profiles for all transcripts in the annotation catalog. 
Note, however, that this is computationally demanding.

```{r}
tx2gene <- system.file("extdata/tx2gene.sub.rds", package = "jcc")
tx2gene <- readRDS(tx2gene)
head(tx2gene)
preds <- predictTxCoverage(biasModel = biasMod$biasModel, 
                           exonsByTx = biasMod$exonsByTx, 
                           bam = bam, tx2Gene = tx2gene, genome = Hsapiens,
                           genes = c("ENSG00000070371", "ENSG00000244296"), 
                           nCores = 1, verbose = TRUE)
preds[2]
```

# Scale transcript coverage profiles

Based on the predicted transcript coverage profiles obtained above and the estimated transcript abundances, we next determine the number of reads predicted to map across each annotated splice junction. 

```{r}
txQuants <- readRDS(system.file("extdata/quant.sub.rds", package = "jcc"))
txsc <- scaleTxCoverages(txCoverageProfiles = preds, 
                         txQuants = txQuants, tx2Gene = tx2gene,
                         strandSpecific = TRUE, methodName = "Salmon", 
                         verbose = TRUE)
names(txsc)
head(txsc$allcovs)
head(txsc$quants)
```

# Combine observed predicted junction coverages

Assuming the reads were aligned to the genome with STAR, we have a text file (with name ending in `SJ.out.tab`) containing the number of reads spanning each annotated junction. 
Below, we read this file, add column names and combine these observed junction coverages with the predicted ones obtained above. 

```{r}
jcov <- read.delim(system.file("extdata/sub.SJ.out.tab", package = "jcc"),
                   header = FALSE, as.is = TRUE) %>%
  setNames(c("seqnames", "start", "end", "strand", "motif", "annot",
             "uniqreads", "mmreads", "maxoverhang")) %>% 
  dplyr::mutate(strand = replace(strand, strand == 1, "+")) %>%
  dplyr::mutate(strand = replace(strand, strand == 2, "-")) %>%
  dplyr::select(seqnames, start, end, strand, uniqreads, mmreads) %>%
  dplyr::mutate(seqnames = as.character(seqnames))

combCov <- combineCoverages(junctionCounts = jcov, 
                            predictedCoverages = txsc$allcovs,
                            txQuants = txsc$quants)
names(combCov)
head(combCov$junctions %>% as.data.frame())
head(combCov$transcripts %>% as.data.frame())
head(combCov$genes %>% as.data.frame())
```

# Calculate JCC scores

Finally, we calculate scaled predicted junction coverages and summarize the differences between the observed and predicted junction coverages for each gene in the JCC (junction coverage compatibility) scores. 

```{r}
jcc <- calculateJCCScores(junctionCounts = combCov$junctions, 
                          txQuantsGene = combCov$genes)
head(jcc$junctions %>% as.data.frame())
head(jcc$genes %>% as.data.frame())
```

# Session info

```{r}
sessionInfo()
```

# References
