<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Combine observed and predicted coverages — combineCoverages • jcc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Combine observed and predicted coverages — combineCoverages" />
<meta property="og:description" content="This function combines predicted junction coverages (from
scaleTxCoverages) and observed junction coverages (obtained
from aligning the reads to the genome and counting the number of reads
spanning each junction). It also summarized transcript abundances on the gene
level and adds information about the number/fraction of uniquely mapping
reads and multimapping reads spanning each junction." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jcc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/JCCscore.html">Calculating JCC scores</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Combine observed and predicted coverages</h1>
    
    <div class="hidden name"><code>combineCoverages.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function combines predicted junction coverages (from
<code><a href='scaleTxCoverages.html'>scaleTxCoverages</a></code>) and observed junction coverages (obtained
from aligning the reads to the genome and counting the number of reads
spanning each junction). It also summarized transcript abundances on the gene
level and adds information about the number/fraction of uniquely mapping
reads and multimapping reads spanning each junction.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>combineCoverages</span><span class='op'>(</span><span class='va'>junctionCounts</span>, <span class='va'>junctionPredCovs</span>, <span class='va'>txQuants</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>junctionCounts</th>
      <td><p>A <code>data.frame</code> with observed junction coverages.
Should have columns <code>seqnames</code>, <code>start</code>, <code>end</code>,
<code>strand</code>, <code>uniqreads</code> and <code>mmreads</code>.</p></td>
    </tr>
    <tr>
      <th>junctionPredCovs</th>
      <td><p>A <code>data.frame</code> with predicted junction
coverages.</p></td>
    </tr>
    <tr>
      <th>txQuants</th>
      <td><p>A <code>data.frame</code> with estimated transcript abundances.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with three elements:</p><dl>
<dt><code>junctionCovs</code>:</dt><dd><p>A
  <code>GRanges</code> object with predicted and observed junction coverages.</p></dd>
  <dt><code>txQuants</code>:</dt><dd><p>A <code>tibble</code> with estimated transcript
  abundances.</p></dd> <dt><code>geneQuants</code>:</dt><dd><p>A <code>tibble</code> with summarized gene
  abundances.</p></dd> 
</dl>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD:
A junction coverage compatibility score to quantify the reliability of
transcript abundance estimates and annotation catalogs. bioRxiv
doi:10.1101/378539 (2018)</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Charlotte Soneson</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span></span>
<span class='r-in'><span class='va'>gtf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz"</span>,</span>
<span class='r-in'>                   package <span class='op'>=</span> <span class='st'>"jcc"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>bam</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/reads.chr22.bam"</span>, package <span class='op'>=</span> <span class='st'>"jcc"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>biasMod</span> <span class='op'>&lt;-</span> <span class='fu'><a href='fitAlpineBiasModel.html'>fitAlpineBiasModel</a></span><span class='op'>(</span>gtf <span class='op'>=</span> <span class='va'>gtf</span>, bam <span class='op'>=</span> <span class='va'>bam</span>,</span>
<span class='r-in'>                              organism <span class='op'>=</span> <span class='st'>"Homo_sapiens"</span>,</span>
<span class='r-in'>                              genome <span class='op'>=</span> <span class='va'>Hsapiens</span>, genomeVersion <span class='op'>=</span> <span class='st'>"GRCh38"</span>,</span>
<span class='r-in'>                              version <span class='op'>=</span> <span class='fl'>90</span>, minLength <span class='op'>=</span> <span class='fl'>230</span>,</span>
<span class='r-in'>                              maxLength <span class='op'>=</span> <span class='fl'>7000</span>, minCount <span class='op'>=</span> <span class='fl'>10</span>,</span>
<span class='r-in'>                              maxCount <span class='op'>=</span> <span class='fl'>10000</span>, subsample <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span class='r-in'>                              nbrSubsample <span class='op'>=</span> <span class='fl'>30</span>, seed <span class='op'>=</span> <span class='fl'>1</span>, minSize <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span class='r-in'>                              maxSize <span class='op'>=</span> <span class='fl'>220</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>tx2gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/tx2gene.sub.rds"</span>, package <span class='op'>=</span> <span class='st'>"jcc"</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>predCovProfiles</span> <span class='op'>&lt;-</span> <span class='fu'><a href='predictTxCoverage.html'>predictTxCoverage</a></span><span class='op'>(</span>biasModel <span class='op'>=</span> <span class='va'>biasMod</span><span class='op'>$</span><span class='va'>biasModel</span>,</span>
<span class='r-in'>                                     exonsByTx <span class='op'>=</span> <span class='va'>biasMod</span><span class='op'>$</span><span class='va'>exonsByTx</span>,</span>
<span class='r-in'>                                     bam <span class='op'>=</span> <span class='va'>bam</span>, tx2gene <span class='op'>=</span> <span class='va'>tx2gene</span>,</span>
<span class='r-in'>                                     genome <span class='op'>=</span> <span class='va'>Hsapiens</span>,</span>
<span class='r-in'>                                     genes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ENSG00000070371"</span>,</span>
<span class='r-in'>                                               <span class='st'>"ENSG00000093010"</span><span class='op'>)</span>,</span>
<span class='r-in'>                                     nCores <span class='op'>=</span> <span class='fl'>1</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>txQuants</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/quant.sub.rds"</span>, package <span class='op'>=</span> <span class='st'>"jcc"</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>txsc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='scaleTxCoverages.html'>scaleTxCoverages</a></span><span class='op'>(</span>txCoverageProfiles <span class='op'>=</span> <span class='va'>predCovProfiles</span>,</span>
<span class='r-in'>                         txQuants <span class='op'>=</span> <span class='va'>txQuants</span>, tx2gene <span class='op'>=</span> <span class='va'>tx2gene</span>,</span>
<span class='r-in'>                         strandSpecific <span class='op'>=</span> <span class='cn'>TRUE</span>, methodName <span class='op'>=</span> <span class='st'>"Salmon"</span>,</span>
<span class='r-in'>                         verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>jcov</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.delim</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/sub.SJ.out.tab"</span>, package <span class='op'>=</span> <span class='st'>"jcc"</span><span class='op'>)</span>,</span>
<span class='r-in'>                   header <span class='op'>=</span> <span class='cn'>FALSE</span>, as.is <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"seqnames"</span>, <span class='st'>"start"</span>, <span class='st'>"end"</span>, <span class='st'>"strand"</span>, <span class='st'>"motif"</span>, <span class='st'>"annot"</span>,</span>
<span class='r-in'>           <span class='st'>"uniqreads"</span>, <span class='st'>"mmreads"</span>, <span class='st'>"maxoverhang"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>strand <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/replace.html'>replace</a></span><span class='op'>(</span><span class='va'>strand</span>, <span class='va'>strand</span> <span class='op'>==</span> <span class='fl'>1</span>, <span class='st'>"+"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>strand <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/replace.html'>replace</a></span><span class='op'>(</span><span class='va'>strand</span>, <span class='va'>strand</span> <span class='op'>==</span> <span class='fl'>2</span>, <span class='st'>"-"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>seqnames</span>, <span class='va'>start</span>, <span class='va'>end</span>, <span class='va'>strand</span>, <span class='va'>uniqreads</span>, <span class='va'>mmreads</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>seqnames <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='va'>seqnames</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>combCov</span> <span class='op'>&lt;-</span> <span class='fu'>combineCoverages</span><span class='op'>(</span>junctionCounts <span class='op'>=</span> <span class='va'>jcov</span>,</span>
<span class='r-in'>                            junctionPredCovs <span class='op'>=</span> <span class='va'>txsc</span><span class='op'>$</span><span class='va'>junctionPredCovs</span>,</span>
<span class='r-in'>                            txQuants <span class='op'>=</span> <span class='va'>txsc</span><span class='op'>$</span><span class='va'>txQuants</span><span class='op'>)</span></span>
<span class='r-in'><span class='op'>}</span></span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Charlotte Soneson.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


