<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fit fragment bias model with alpine — fitAlpineBiasModel • jcc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fit fragment bias model with alpine — fitAlpineBiasModel" />
<meta property="og:description" content="This function provides a wrapper around some of the functions from the
alpine package. Given a gtf file and a bam file with reads aligned to
the genome, it will find single-isoform genes (with lengths and expression
levels within given ranges) and use the observed read coverages to fit a
fragment bias model." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jcc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/JCCscore.html">Calculating JCC scores</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit fragment bias model with alpine</h1>
    
    <div class="hidden name"><code>fitAlpineBiasModel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function provides a wrapper around some of the functions from the
<code>alpine</code> package. Given a gtf file and a bam file with reads aligned to
the genome, it will find single-isoform genes (with lengths and expression
levels within given ranges) and use the observed read coverages to fit a
fragment bias model.</p>
    </div>

    <pre class="usage"><span class='fu'>fitAlpineBiasModel</span>(<span class='no'>gtf</span>, <span class='no'>bam</span>, <span class='no'>organism</span>, <span class='no'>genome</span>, <span class='no'>genomeVersion</span>, <span class='no'>version</span>,
  <span class='kw'>minLength</span> <span class='kw'>=</span> <span class='fl'>600</span>, <span class='kw'>maxLength</span> <span class='kw'>=</span> <span class='fl'>7000</span>, <span class='kw'>minCount</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>maxCount</span> <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>subsample</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>nbrSubsample</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>minSize</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>maxSize</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gtf</th>
      <td><p>Path to gtf file with genomic features. Preferably in Ensembl
format.</p></td>
    </tr>
    <tr>
      <th>bam</th>
      <td><p>Path to bam file with read alignments to the genome.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>The organism (e.g., 'Homo_sapiens'). This argument will be
passed to <code><a href='https://rdrr.io/pkg/ensembldb/man/makeEnsemblDbPackage.html'>ensembldb::ensDbFromGtf</a></code>.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>A <code>BSgenome</code> object.</p></td>
    </tr>
    <tr>
      <th>genomeVersion</th>
      <td><p>Genome version (e.g., 'GRCh38'). This argument will be
passed to <code><a href='https://rdrr.io/pkg/ensembldb/man/makeEnsemblDbPackage.html'>ensembldb::ensDbFromGtf</a></code>.</p></td>
    </tr>
    <tr>
      <th>version</th>
      <td><p>The version of the reference annotation (e.g., 90). This
argument will be passed to <code><a href='https://rdrr.io/pkg/ensembldb/man/makeEnsemblDbPackage.html'>ensembldb::ensDbFromGtf</a></code>.</p></td>
    </tr>
    <tr>
      <th>minLength, maxLength</th>
      <td><p>Minimum and maximum length of single-isoform genes
used to fit fragment bias model.</p></td>
    </tr>
    <tr>
      <th>minCount, maxCount</th>
      <td><p>Minimum and maximum read coverage of single-isoform
genes used to fit fragment bias model.</p></td>
    </tr>
    <tr>
      <th>subsample</th>
      <td><p>Whether to subsample the set of single-isoform genes
satisfying the <code>minLength</code>, <code>maxLength</code>, <code>minCount</code> and
<code>maxCount</code> criteria before fitting the fragment bias model.</p></td>
    </tr>
    <tr>
      <th>nbrSubsample</th>
      <td><p>If <code>subsample</code> is <code>TRUE</code>, the number of genes
to subsample.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>If <code>subsample</code> is <code>TRUE</code>, the random seed to use to
ensure reproducibility.</p></td>
    </tr>
    <tr>
      <th>minSize, maxSize</th>
      <td><p>Smallest and largest fragment size to consider. One or
both of these can be <code>NULL</code>, in which case it is estimated as the 2.5
or 97.5 percentile, respectively, of estimated fragment sizes in the
provided data.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical, whether to print progress messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with three elements:</p><dl>
<dt><code>biasModel</code>:</dt><dd><p>The
  fitted fragment bias model.</p></dd> <dt><code>exonsByTx</code>:</dt><dd><p>A <code>GRangesList</code>
  object with exons grouped by transcript.</p></dd> <dt><code>transcripts</code>:</dt><dd><p>A
  <code>GRanges</code> object with all the reference transcripts.</p></dd> 
</dl>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD:
A junction coverage compatibility score to quantify the reliability of
transcript abundance estimates and annotation catalogs. bioRxiv
doi:10.1101/378539 (2018).</p>
<p>Love MI, Hogenesch JB, Irizarry RA: Modeling of RNA-seq fragment sequence
bias reduces systematic errors in transcript abundance estimation. Nature
Biotechnology 34(12):1287-1291 (2016).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>gtf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz"</span>,
                   <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>)
<span class='no'>bam</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/reads.chr22.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>)
<span class='no'>biasMod</span> <span class='kw'>&lt;-</span> <span class='fu'>fitAlpineBiasModel</span>(<span class='kw'>gtf</span> <span class='kw'>=</span> <span class='no'>gtf</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bam</span>,
                              <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo_sapiens"</span>,
                              <span class='kw'>genome</span> <span class='kw'>=</span> <span class='no'>Hsapiens</span>, <span class='kw'>genomeVersion</span> <span class='kw'>=</span> <span class='st'>"GRCh38"</span>,
                              <span class='kw'>version</span> <span class='kw'>=</span> <span class='fl'>90</span>, <span class='kw'>minLength</span> <span class='kw'>=</span> <span class='fl'>230</span>,
                              <span class='kw'>maxLength</span> <span class='kw'>=</span> <span class='fl'>7000</span>, <span class='kw'>minCount</span> <span class='kw'>=</span> <span class='fl'>10</span>,
                              <span class='kw'>maxCount</span> <span class='kw'>=</span> <span class='fl'>10000</span>, <span class='kw'>subsample</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
                              <span class='kw'>nbrSubsample</span> <span class='kw'>=</span> <span class='fl'>30</span>, <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>minSize</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                              <span class='kw'>maxSize</span> <span class='kw'>=</span> <span class='fl'>220</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Charlotte Soneson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


