<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot JCC score summary for a given gene — plotGeneSummary • jcc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot JCC score summary for a given gene — plotGeneSummary" />
<meta property="og:description" content="This function generates a multi-panel plot, including (i) the observed
coverage and gene model, (ii) the estimated transcript abundances and (iii)
the observed and predicted junction coverages within the gene." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jcc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/JCCscore.html">Calculating JCC scores</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot JCC score summary for a given gene</h1>
    
    <div class="hidden name"><code>plotGeneSummary.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function generates a multi-panel plot, including (i) the observed
coverage and gene model, (ii) the estimated transcript abundances and (iii)
the observed and predicted junction coverages within the gene.</p>
    </div>

    <pre class="usage"><span class='fu'>plotGeneSummary</span>(<span class='no'>gtf</span>, <span class='no'>junctionCovs</span>, <span class='no'>useGene</span>, <span class='no'>bwFile</span>, <span class='no'>txQuants</span>,
  <span class='kw'>txCol</span> <span class='kw'>=</span> <span class='st'>"transcript_id"</span>, <span class='kw'>geneCol</span> <span class='kw'>=</span> <span class='st'>"gene_id"</span>, <span class='kw'>exonCol</span> <span class='kw'>=</span> <span class='st'>"exon_id"</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gtf</th>
      <td><p>Path to gtf file with genomic features. Preferably in Ensembl
format.</p></td>
    </tr>
    <tr>
      <th>junctionCovs</th>
      <td><p>A <code>data.frame</code> with junction coverage information,
output from <code><a href='calculateJCCScores.html'>calculateJCCScores</a></code>.</p></td>
    </tr>
    <tr>
      <th>useGene</th>
      <td><p>A character string giving the name of the gene to plot.</p></td>
    </tr>
    <tr>
      <th>bwFile</th>
      <td><p>Path to a bigwig file generated from a genome alignment BAM
file.</p></td>
    </tr>
    <tr>
      <th>txQuants</th>
      <td><p>A <code>data.frame</code> with estimated transcript abundances,
output from <code><a href='combineCoverages.html'>combineCoverages</a></code>.</p></td>
    </tr>
    <tr>
      <th>txCol</th>
      <td><p>The column in the gtf file that contains the transcript ID.</p></td>
    </tr>
    <tr>
      <th>geneCol</th>
      <td><p>The column in the gtf file that contains the gene ID.</p></td>
    </tr>
    <tr>
      <th>exonCol</th>
      <td><p>The column in the gtf file that contains the exon ID.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>ggplot</code> object with a multi-panel plot. The size is optimized
  for display in a device of size approximately 12in wide and 10in high.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD: A junction
coverage compatibility score to quantify the reliability of transcript
abundance estimates and annotation catalogs. bioRxiv doi:10.1101/378539
(2018)</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>gtf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz"</span>,
                   <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>)
<span class='no'>bam</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/reads.chr22.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>)
<span class='no'>biasMod</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='fitAlpineBiasModel.html'>fitAlpineBiasModel</a></span>(<span class='kw'>gtf</span> <span class='kw'>=</span> <span class='no'>gtf</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bam</span>,
                              <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo_sapiens"</span>,
                              <span class='kw'>genome</span> <span class='kw'>=</span> <span class='no'>Hsapiens</span>, <span class='kw'>genomeVersion</span> <span class='kw'>=</span> <span class='st'>"GRCh38"</span>,
                              <span class='kw'>version</span> <span class='kw'>=</span> <span class='fl'>90</span>, <span class='kw'>minLength</span> <span class='kw'>=</span> <span class='fl'>230</span>,
                              <span class='kw'>maxLength</span> <span class='kw'>=</span> <span class='fl'>7000</span>, <span class='kw'>minCount</span> <span class='kw'>=</span> <span class='fl'>10</span>,
                              <span class='kw'>maxCount</span> <span class='kw'>=</span> <span class='fl'>10000</span>, <span class='kw'>subsample</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
                              <span class='kw'>nbrSubsample</span> <span class='kw'>=</span> <span class='fl'>30</span>, <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>minSize</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                              <span class='kw'>maxSize</span> <span class='kw'>=</span> <span class='fl'>220</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>tx2gene</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/tx2gene.sub.rds"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>))
<span class='no'>predCovProfiles</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='predictTxCoverage.html'>predictTxCoverage</a></span>(<span class='kw'>biasModel</span> <span class='kw'>=</span> <span class='no'>biasMod</span>$<span class='no'>biasModel</span>,
                                     <span class='kw'>exonsByTx</span> <span class='kw'>=</span> <span class='no'>biasMod</span>$<span class='no'>exonsByTx</span>,
                                     <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bam</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='no'>tx2gene</span>,
                                     <span class='kw'>genome</span> <span class='kw'>=</span> <span class='no'>Hsapiens</span>,
                                     <span class='kw'>genes</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000070371"</span>,
                                               <span class='st'>"ENSG00000093010"</span>),
                                     <span class='kw'>nCores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>txQuants</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/quant.sub.rds"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>))
<span class='no'>txsc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='scaleTxCoverages.html'>scaleTxCoverages</a></span>(<span class='kw'>txCoverageProfiles</span> <span class='kw'>=</span> <span class='no'>predCovProfiles</span>,
                         <span class='kw'>txQuants</span> <span class='kw'>=</span> <span class='no'>txQuants</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='no'>tx2gene</span>,
                         <span class='kw'>strandSpecific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>methodName</span> <span class='kw'>=</span> <span class='st'>"Salmon"</span>,
                         <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>jcov</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.delim</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/sub.SJ.out.tab"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>),
                   <span class='kw'>header</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>as.is</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='kw'>%&gt;%</span>
 <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"seqnames"</span>, <span class='st'>"start"</span>, <span class='st'>"end"</span>, <span class='st'>"strand"</span>, <span class='st'>"motif"</span>, <span class='st'>"annot"</span>,
           <span class='st'>"uniqreads"</span>, <span class='st'>"mmreads"</span>, <span class='st'>"maxoverhang"</span>)) <span class='kw'>%&gt;%</span>
 <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span>(<span class='kw'>strand</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/replace.html'>replace</a></span>(<span class='no'>strand</span>, <span class='no'>strand</span> <span class='kw'>==</span> <span class='fl'>1</span>, <span class='st'>"+"</span>)) <span class='kw'>%&gt;%</span>
 <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span>(<span class='kw'>strand</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/replace.html'>replace</a></span>(<span class='no'>strand</span>, <span class='no'>strand</span> <span class='kw'>==</span> <span class='fl'>2</span>, <span class='st'>"-"</span>)) <span class='kw'>%&gt;%</span>
 <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span>(<span class='no'>seqnames</span>, <span class='no'>start</span>, <span class='no'>end</span>, <span class='no'>strand</span>, <span class='no'>uniqreads</span>, <span class='no'>mmreads</span>) <span class='kw'>%&gt;%</span>
 <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span>(<span class='kw'>seqnames</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span>(<span class='no'>seqnames</span>))
<span class='no'>combCov</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='combineCoverages.html'>combineCoverages</a></span>(<span class='kw'>junctionCounts</span> <span class='kw'>=</span> <span class='no'>jcov</span>,
                            <span class='kw'>junctionPredCovs</span> <span class='kw'>=</span> <span class='no'>txsc</span>$<span class='no'>junctionPredCovs</span>,
                            <span class='kw'>txQuants</span> <span class='kw'>=</span> <span class='no'>txsc</span>$<span class='no'>txQuants</span>)
<span class='no'>jcc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='calculateJCCScores.html'>calculateJCCScores</a></span>(<span class='kw'>junctionCovs</span> <span class='kw'>=</span> <span class='no'>combCov</span>$<span class='no'>junctionCovs</span>,
                          <span class='kw'>geneQuants</span> <span class='kw'>=</span> <span class='no'>combCov</span>$<span class='no'>geneQuants</span>)
<span class='no'>bwFile</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/reads.chr22.bw"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"jcc"</span>)
<span class='fu'>plotGeneSummary</span>(<span class='kw'>gtf</span> <span class='kw'>=</span> <span class='no'>gtf</span>, <span class='kw'>junctionCovs</span> <span class='kw'>=</span> <span class='no'>jcc</span>$<span class='no'>junctionCovs</span>,
                <span class='kw'>useGene</span> <span class='kw'>=</span> <span class='st'>"ENSG00000070371"</span>, <span class='kw'>bwFile</span> <span class='kw'>=</span> <span class='no'>bwFile</span>,
                <span class='kw'>txQuants</span> <span class='kw'>=</span> <span class='no'>combCov</span>$<span class='no'>txQuants</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Charlotte Soneson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


