<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Scale transcript coverages by abundance estimate — scaleTxCoverages • jcc</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Scale transcript coverages by abundance estimate — scaleTxCoverages"><meta property="og:description" content="This function takes a list of predicted transcript coverage profiles (from
predictTxCoverage) and corresponding transcript abundance
estimates (from, e.g., Salmon, kallisto, StringTie, RSEM, or any other
transcript abundance estimation software), and scales the coverage profiles
by the abundance estimates in order to determine the predicted number of
reads covering each position in the transcript. In addition, it extracts the
predicted number of reads spanning each junction in each transcript, and sums
up these for each annotated junction, across all transcripts containing the
junction."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jcc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/JCCscore.html">Calculating JCC scores</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Scale transcript coverages by abundance estimate</h1>

    <div class="hidden name"><code>scaleTxCoverages.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes a list of predicted transcript coverage profiles (from
<code><a href="predictTxCoverage.html">predictTxCoverage</a></code>) and corresponding transcript abundance
estimates (from, e.g., Salmon, kallisto, StringTie, RSEM, or any other
transcript abundance estimation software), and scales the coverage profiles
by the abundance estimates in order to determine the predicted number of
reads covering each position in the transcript. In addition, it extracts the
predicted number of reads spanning each junction in each transcript, and sums
up these for each annotated junction, across all transcripts containing the
junction.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scaleTxCoverages</span><span class="op">(</span><span class="va">txCoverageProfiles</span>, <span class="va">txQuants</span>, <span class="va">tx2gene</span>, <span class="va">strandSpecific</span>,</span>
<span>  <span class="va">methodName</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-txcoverageprofiles">txCoverageProfiles<a class="anchor" aria-label="anchor" href="#arg-txcoverageprofiles"></a></dt>
<dd><p>A <code>list</code> of predicted transcript coverage
profiles (output from <code><a href="predictTxCoverage.html">predictTxCoverage</a></code>).</p></dd>


<dt id="arg-txquants">txQuants<a class="anchor" aria-label="anchor" href="#arg-txquants"></a></dt>
<dd><p>A <code>data.frame</code> with estimated transcript abundances.
Must contain columns named <code>transcript</code>, <code>TPM</code> and <code>count</code>,
all other columns will be ignored.</p></dd>


<dt id="arg-tx-gene">tx2gene<a class="anchor" aria-label="anchor" href="#arg-tx-gene"></a></dt>
<dd><p>A <code>data.frame</code> with transcript-to-gene mapping. Must have
at least two columns, named <code>tx</code> and <code>gene</code>.</p></dd>


<dt id="arg-strandspecific">strandSpecific<a class="anchor" aria-label="anchor" href="#arg-strandspecific"></a></dt>
<dd><p>Logical, whether or not the data is strand-specific.
Determines whether or not the strand of a junction should be taken into
account when the coverages are summed up across transcripts.</p></dd>


<dt id="arg-methodname">methodName<a class="anchor" aria-label="anchor" href="#arg-methodname"></a></dt>
<dd><p>Name to use for the abundance estimation method in the
output table.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, whether to print progress messages.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with two elements:</p><dl><dt><code>junctionPredCovs</code>:</dt>
<dd><p>A <code>tibble</code> with the predicted coverage
for each junction, scaled by the estimated transcript abundances and
summarized across all transcripts.</p></dd>

<dt><code>txQuants</code>:</dt>
<dd><p>A <code>tibble</code> with estimated transcript abundances,
including information about whether or not the coverage profile could be
predicted ('covOK') or if a uniform coverage was imposed ('covError' or
'covNA').</p></dd>


</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Soneson C, Love MI, Patro R, Hussain S, Malhotra D, Robinson MD: A junction
coverage compatibility score to quantify the reliability of transcript
abundance estimates and annotation catalogs. bioRxiv doi:10.1101/378539
(2018)</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Charlotte Soneson</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz"</span>,</span></span>
<span class="r-in"><span>                   package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/reads.chr22.bam"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">biasMod</span> <span class="op">&lt;-</span> <span class="fu"><a href="fitAlpineBiasModel.html">fitAlpineBiasModel</a></span><span class="op">(</span>gtf <span class="op">=</span> <span class="va">gtf</span>, bam <span class="op">=</span> <span class="va">bam</span>,</span></span>
<span class="r-in"><span>                              organism <span class="op">=</span> <span class="st">"Homo_sapiens"</span>,</span></span>
<span class="r-in"><span>                              genome <span class="op">=</span> <span class="va">Hsapiens</span>, genomeVersion <span class="op">=</span> <span class="st">"GRCh38"</span>,</span></span>
<span class="r-in"><span>                              version <span class="op">=</span> <span class="fl">90</span>, minLength <span class="op">=</span> <span class="fl">230</span>,</span></span>
<span class="r-in"><span>                              maxLength <span class="op">=</span> <span class="fl">7000</span>, minCount <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                              maxCount <span class="op">=</span> <span class="fl">10000</span>, subsample <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                              nbrSubsample <span class="op">=</span> <span class="fl">30</span>, seed <span class="op">=</span> <span class="fl">1</span>, minSize <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                              maxSize <span class="op">=</span> <span class="fl">220</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tx2gene.sub.rds"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">predCovProfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="predictTxCoverage.html">predictTxCoverage</a></span><span class="op">(</span>biasModel <span class="op">=</span> <span class="va">biasMod</span><span class="op">$</span><span class="va">biasModel</span>,</span></span>
<span class="r-in"><span>                                     exonsByTx <span class="op">=</span> <span class="va">biasMod</span><span class="op">$</span><span class="va">exonsByTx</span>,</span></span>
<span class="r-in"><span>                                     bam <span class="op">=</span> <span class="va">bam</span>, tx2gene <span class="op">=</span> <span class="va">tx2gene</span>,</span></span>
<span class="r-in"><span>                                     genome <span class="op">=</span> <span class="va">Hsapiens</span>,</span></span>
<span class="r-in"><span>                                     genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000070371"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"ENSG00000093010"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                     nCores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">txQuants</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/quant.sub.rds"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">txsc</span> <span class="op">&lt;-</span> <span class="fu">scaleTxCoverages</span><span class="op">(</span>txCoverageProfiles <span class="op">=</span> <span class="va">predCovProfiles</span>,</span></span>
<span class="r-in"><span>                         txQuants <span class="op">=</span> <span class="va">txQuants</span>, tx2gene <span class="op">=</span> <span class="va">tx2gene</span>,</span></span>
<span class="r-in"><span>                         strandSpecific <span class="op">=</span> <span class="cn">TRUE</span>, methodName <span class="op">=</span> <span class="st">"Salmon"</span>,</span></span>
<span class="r-in"><span>                         verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Charlotte Soneson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

      </footer></div>






  </body></html>

