<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Calculating JCC scores • jcc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Calculating JCC scores">
<meta property="og:description" content="jcc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jcc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/JCCscore.html">Calculating JCC scores</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="JCCscore_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Calculating JCC scores</h1>
                        <h4 data-toc-skip class="author">Charlotte Soneson</h4>
                                    <h4 data-toc-skip class="author">Michael I Love</h4>
                                    <h4 data-toc-skip class="author">Rob Patro</h4>
                                    <h4 data-toc-skip class="author">Shobbir Hussain</h4>
                                    <h4 data-toc-skip class="author">Dheeraj Malhotra</h4>
                                    <h4 data-toc-skip class="author">Mark D Robinson</h4>
                        
            <h4 data-toc-skip class="date">2021-10-30</h4>
      
      
      <div class="hidden name"><code>JCCscore.Rmd</code></div>

    </div>

    
    
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>This vignette describes how to use the <code>jcc</code> package to calculate junction coverage compatibility (JCC_ scores <span class="citation">(Soneson et al. 2018)</span> for a given set of genes. In order to run the entire workflow, the following files are required, and will be used as described below:</p>
<ul>
<li>A BAM file with reads aligned to a reference genome. The package has been tested with BAM files generated by the STAR aligner <span class="citation">(Dobin et al. 2013)</span>, but files generated by other genome aligners should also work. The reads in the BAM file will be used to fit the fragment bias model with <code>r Biocpkg("alpine")</code> <span class="citation">(Love, Hogenesch, and Irizarry 2016)</span> as well as to obtain the predicted transcript coverage profiles.</li>
<li>A <code>data.frame</code> with the number of uniquely mapping and multimapping reads aligned across each annotated splice junction. Below we show how to generate a correctly formatted data frame from the <code>SJ.out.tab</code> text file generated by STAR.</li>
<li>A <code>BSgenome</code> object for the genome that the reads in the BAM file were aligned to.</li>
<li>A <code>gtf</code> file corresponding to the <code>BSgenome</code> and the BAM file.</li>
<li>A transcript-to-gene conversion <code>data.frame</code> with at least two columns, named <code>tx</code> and <code>gene</code>. Additional columns providing, e.g., gene symbols can be included.</li>
<li>A <code>data.frame</code> with estimated transcript abundances, with three columns named <code>transcript</code>, <code>TPM</code> and <code>count</code>.</li>
<li>If you want to create gene summary plots, including an illustration of the observed coverage in a genomic region, you also need to convert the BAM file into a bigWig file. This can be done e.g. as follows:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="ex">bedtools</span> genomecov -split -ibam <span class="op">&lt;</span>file.bam<span class="op">&gt;</span> -bg <span class="op">&gt;</span> <span class="op">&lt;</span>file.bedGraph<span class="op">&gt;</span> </span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="ex">bedGraphToBigWig</span> <span class="op">&lt;</span>file.bedGraph<span class="op">&gt;</span> <span class="op">&lt;</span>chrNameLength.txt<span class="op">&gt;</span> <span class="op">&lt;</span>file.bw<span class="op">&gt;</span></span></code></pre></div>
<p>where you should replace the text in <code>&lt;&gt;</code> by the appropriate file names. The <code>&lt;chrNameLength.txt&gt;</code> is a text file with the length of each chromosome. If you created a genome index with STAR, the index directory contains such a file. The information can also be extracted from the header of the BAM file.</p>
<p>The workflow has mainly been tested with reference files from Ensembl, and may need to be adjusted to work with reference files in other formats.</p>
<p>The general procedure for estimating junction coverage compatibility (JCC) scores for a set of genes is outlined in the figure below. Please refer to the paper <span class="citation">(Soneson et al. 2018)</span> for a more detailed description and motivation of the different steps.</p>
<p><img src="study_design.png" alt="drawing" width="400px"></p>
<p>In the rest of this vignette, we will illustrate each step using a small example data set provided with the package.</p>
</div>
<div class="section level1">
<h1 id="fit-fragment-bias-model">Fit fragment bias model<a class="anchor" aria-label="anchor" href="#fit-fragment-bias-model"></a>
</h1>
<p>The first step in the process is to fit a fragment bias model using the <em><a href="https://bioconductor.org/packages/3.15/alpine" class="external-link">alpine</a></em> package. The <code><a href="../reference/fitAlpineBiasModel.html">fitAlpineBiasModel()</a></code> function provides a wrapper containing the necessary steps. Note that the parameters of this function may need to be adapted to be appropriate for a specific data set. In particular, since the example data set only contains a small number of reads and genes, we set <code>minCount</code> and <code>minLength</code> to small values to make sure that some genes will be retained and used to fit the fragment bias model. However, in a real data set these arguments can often be set to higher values. Also, the <code>BSgenome</code> object (here, <code>Hsapiens</code> from the <code>BSgenome.Hsapiens.NCBI.GRCh38</code> package) should be replaced with the appropriate object for your organism. We refer to the help file and the <em><a href="https://bioconductor.org/packages/3.15/alpine" class="external-link">alpine</a></em> package for more detailed explanations of each parameter.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load necessary packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">jcc</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load gtf file and bam file with aligned reads</span>
<span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/Homo_sapiens.GRCh38.90.chr22.gtf.gz"</span>, 
                   package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span>
<span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/reads.chr22.bam"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span>

<span class="co">## Fit fragment bias model</span>
<span class="va">biasMod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitAlpineBiasModel.html">fitAlpineBiasModel</a></span><span class="op">(</span>gtf <span class="op">=</span> <span class="va">gtf</span>, bam <span class="op">=</span> <span class="va">bam</span>, organism <span class="op">=</span> <span class="st">"Homo_sapiens"</span>,
                              genome <span class="op">=</span> <span class="va">Hsapiens</span>, genomeVersion <span class="op">=</span> <span class="st">"GRCh38"</span>,
                              version <span class="op">=</span> <span class="fl">90</span>, minLength <span class="op">=</span> <span class="fl">230</span>, maxLength <span class="op">=</span> <span class="fl">7000</span>,
                              minCount <span class="op">=</span> <span class="fl">10</span>, maxCount <span class="op">=</span> <span class="fl">10000</span>, 
                              subsample <span class="op">=</span> <span class="cn">TRUE</span>, nbrSubsample <span class="op">=</span> <span class="fl">30</span>, seed <span class="op">=</span> <span class="fl">1</span>,
                              minSize <span class="op">=</span> <span class="cn">NULL</span>, maxSize <span class="op">=</span> <span class="fl">220</span>, 
                              verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Creating TxDb object...</span>
<span class="co">#&gt; Importing GTF file ... OK</span>
<span class="co">#&gt; Processing metadata ... OK</span>
<span class="co">#&gt; Processing genes ...</span>
<span class="co">#&gt; Warning in ensDbFromGRanges(GTF, outfile = outfile, path = path, organism</span>
<span class="co">#&gt; = organism, : I'm missing column(s): 'entrezid'. The corresponding database</span>
<span class="co">#&gt; column(s) will be empty!</span>
<span class="co">#&gt;  Attribute availability:</span>
<span class="co">#&gt;   o gene_id ... OK</span>
<span class="co">#&gt;   o gene_name ... OK</span>
<span class="co">#&gt;   o entrezid ... Nope</span>
<span class="co">#&gt;   o gene_biotype ... OK</span>
<span class="co">#&gt; OK</span>
<span class="co">#&gt; Processing transcripts ... </span>
<span class="co">#&gt;  Attribute availability:</span>
<span class="co">#&gt;   o transcript_id ... OK</span>
<span class="co">#&gt;   o gene_id ... OK</span>
<span class="co">#&gt;   o transcript_biotype ... OK</span>
<span class="co">#&gt; OK</span>
<span class="co">#&gt; Processing exons ... OK</span>
<span class="co">#&gt; Processing chromosomes ... Fetch seqlengths from ensembl ... OK</span>
<span class="co">#&gt; Generating index ... OK</span>
<span class="co">#&gt;   -------------</span>
<span class="co">#&gt; Verifying validity of the information in the database:</span>
<span class="co">#&gt; Checking transcripts ... OK</span>
<span class="co">#&gt; Checking exons ... OK</span>
<span class="co">#&gt; Import genomic features from the file as a GRanges object ... OK</span>
<span class="co">#&gt; Prepare the 'metadata' data frame ... OK</span>
<span class="co">#&gt; Make the TxDb object ...</span>
<span class="co">#&gt; Warning in .get_cds_IDX(mcols0$type, mcols0$phase): The "phase" metadata column contains non-NA values for features of type</span>
<span class="co">#&gt;   stop_codon. This information was ignored.</span>
<span class="co">#&gt; OK</span>
<span class="co">#&gt; Getting list of transcripts...</span>
<span class="co">#&gt; Generating list of exons by transcript...</span>
<span class="co">#&gt; Selecting genes with a single isoform...</span>
<span class="co">#&gt; 149 single-isoform genes found.</span>
<span class="co">#&gt; Filtering out too short and too long transcripts...</span>
<span class="co">#&gt; 125 transcripts remaining.</span>
<span class="co">#&gt; Reading bam file...</span>
<span class="co">#&gt; Subsetting to medium-to-highly expressed genes...</span>
<span class="co">#&gt; 28 genes retained.</span>
<span class="co">#&gt; Subsampling genes for fitting bias model...</span>
<span class="co">#&gt; 28 genes retained.</span>
<span class="co">#&gt; Getting fragment widths and read lengths...</span>
<span class="co">#&gt; Warning in .make_GAlignmentPairs_from_GAlignments(gal, strandMode = strandMode, :   3 alignments with ambiguous pairing were dumped.</span>
<span class="co">#&gt;     Use 'getDumpedAlignments()' to retrieve them from the dump environment.</span>
<span class="co">#&gt; Building fragment types...</span>
<span class="co">#&gt; Fitting bias model...</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span></code></pre></div>
<p>The output of <code>fitAlpineBiasModel</code> is a list with three objects:</p>
<ul>
<li>
<code>biasModel</code>: the fitted fragment bias model from <em><a href="https://bioconductor.org/packages/3.15/alpine" class="external-link">alpine</a></em>.</li>
<li>
<code>exonsByTx</code>: a <code>GRangesList</code>, grouping exons by transcript.</li>
<li>
<code>transcripts</code>: a <code>GRanges</code> object with all reference transcripts.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">biasMod</span><span class="op">)</span>
<span class="co">#&gt; [1] "biasModel"   "exonsByTx"   "transcripts"</span>

<span class="va">biasMod</span><span class="op">$</span><span class="va">exonsByTx</span>
<span class="co">#&gt; GRangesList object of length 510:</span>
<span class="co">#&gt; $ENST00000624155</span>
<span class="co">#&gt; GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="co">#&gt;       seqnames            ranges strand |   exon_id       exon_name exon_rank</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt; &lt;integer&gt;</span>
<span class="co">#&gt;   [1]       22 11066418-11066515      + |         1 ENSE00003758096         1</span>
<span class="co">#&gt;   [2]       22 11067985-11068174      + |         2 ENSE00003758861         2</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ENST00000422332</span>
<span class="co">#&gt; GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="co">#&gt;       seqnames            ranges strand |   exon_id       exon_name exon_rank</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt; &lt;integer&gt;</span>
<span class="co">#&gt;   [1]       22 11124337-11124379      + |         3 ENSE00001616995         1</span>
<span class="co">#&gt;   [2]       22 11124508-11125705      + |         4 ENSE00001776473         2</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ENST00000634617</span>
<span class="co">#&gt; GRanges object with 9 ranges and 3 metadata columns:</span>
<span class="co">#&gt;       seqnames            ranges strand |   exon_id       exon_name exon_rank</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt; &lt;integer&gt;</span>
<span class="co">#&gt;   [1]       22 12602466-12602527      + |         5 ENSE00003784219         1</span>
<span class="co">#&gt;   [2]       22 12604670-12604739      + |         6 ENSE00003784834         2</span>
<span class="co">#&gt;   [3]       22 12615409-12615534      + |         7 ENSE00003785913         3</span>
<span class="co">#&gt;   [4]       22 12616314-12616371      + |         8 ENSE00003790031         4</span>
<span class="co">#&gt;   [5]       22 12618171-12618289      + |         9 ENSE00003786552         5</span>
<span class="co">#&gt;   [6]       22 12620838-12620938      + |        10 ENSE00003788654         6</span>
<span class="co">#&gt;   [7]       22 12624242-12624331      + |        11 ENSE00003791453         7</span>
<span class="co">#&gt;   [8]       22 12625317-12625428      + |        12 ENSE00003788301         8</span>
<span class="co">#&gt;   [9]       22 12626607-12626642      + |        13 ENSE00003788969         9</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ...</span>
<span class="co">#&gt; &lt;507 more elements&gt;</span>

<span class="va">biasMod</span><span class="op">$</span><span class="va">transcripts</span>
<span class="co">#&gt; GRanges object with 510 ranges and 2 metadata columns:</span>
<span class="co">#&gt;                   seqnames            ranges strand |         tx_name</span>
<span class="co">#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;</span>
<span class="co">#&gt;   ENST00000624155       22 11066418-11068174      + | ENST00000624155</span>
<span class="co">#&gt;   ENST00000422332       22 11124337-11125705      + | ENST00000422332</span>
<span class="co">#&gt;   ENST00000634617       22 12602466-12626642      + | ENST00000634617</span>
<span class="co">#&gt;   ENST00000613107       22 15273855-15273961      + | ENST00000613107</span>
<span class="co">#&gt;   ENST00000448473       22 15290718-15297196      + | ENST00000448473</span>
<span class="co">#&gt;               ...      ...               ...    ... .             ...</span>
<span class="co">#&gt;   ENST00000473551       22 19979997-19990792      - | ENST00000473551</span>
<span class="co">#&gt;   ENST00000487793       22 19981152-19987190      - | ENST00000487793</span>
<span class="co">#&gt;   ENST00000492625       22 19981382-19987155      - | ENST00000492625</span>
<span class="co">#&gt;   ENST00000462319       22 19981545-19983753      - | ENST00000462319</span>
<span class="co">#&gt;   ENST00000467828       22 19981991-20016745      - | ENST00000467828</span>
<span class="co">#&gt;                           gene_id</span>
<span class="co">#&gt;                       &lt;character&gt;</span>
<span class="co">#&gt;   ENST00000624155 ENSG00000279973</span>
<span class="co">#&gt;   ENST00000422332 ENSG00000226444</span>
<span class="co">#&gt;   ENST00000634617 ENSG00000283023</span>
<span class="co">#&gt;   ENST00000613107 ENSG00000276138</span>
<span class="co">#&gt;   ENST00000448473 ENSG00000236235</span>
<span class="co">#&gt;               ...             ...</span>
<span class="co">#&gt;   ENST00000473551 ENSG00000099889</span>
<span class="co">#&gt;   ENST00000487793 ENSG00000099889</span>
<span class="co">#&gt;   ENST00000492625 ENSG00000099889</span>
<span class="co">#&gt;   ENST00000462319 ENSG00000099889</span>
<span class="co">#&gt;   ENST00000467828 ENSG00000099889</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="predict-transcript-coverage-profiles">Predict transcript coverage profiles<a class="anchor" aria-label="anchor" href="#predict-transcript-coverage-profiles"></a>
</h1>
<p>Next, we use the fitted fragment bias model to predict the coverage profile for each transcript in a set of specified genes. The <code>genes</code> argument can be set to <code>NULL</code> to predict the coverage profiles for all transcripts in the annotation catalog. Note, however, that this is computationally demanding if the number of transcripts is large. The <code>nCores</code> argument can be increased to predict coverage profiles for several transcripts in parallel.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load transcript-to-gene conversion table</span>
<span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tx2gene.sub.rds"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">)</span>
<span class="co">#&gt;                   tx            gene symbol   gene_biotype      tx_biotype</span>
<span class="co">#&gt; 4524 ENST00000621271 ENSG00000070371 CLTCL1 protein_coding  protein_coding</span>
<span class="co">#&gt; 4525 ENST00000427926 ENSG00000070371 CLTCL1 protein_coding  protein_coding</span>
<span class="co">#&gt; 4526 ENST00000412649 ENSG00000070371 CLTCL1 protein_coding retained_intron</span>
<span class="co">#&gt; 4527 ENST00000622493 ENSG00000070371 CLTCL1 protein_coding  protein_coding</span>
<span class="co">#&gt; 4528 ENST00000617926 ENSG00000070371 CLTCL1 protein_coding  protein_coding</span>
<span class="co">#&gt; 4529 ENST00000615606 ENSG00000070371 CLTCL1 protein_coding retained_intron</span>

<span class="co">## Predict transcript coverage profiles</span>
<span class="va">predCovProfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictTxCoverage.html">predictTxCoverage</a></span><span class="op">(</span>biasModel <span class="op">=</span> <span class="va">biasMod</span><span class="op">$</span><span class="va">biasModel</span>, 
                                     exonsByTx <span class="op">=</span> <span class="va">biasMod</span><span class="op">$</span><span class="va">exonsByTx</span>, 
                                     bam <span class="op">=</span> <span class="va">bam</span>, tx2gene <span class="op">=</span> <span class="va">tx2gene</span>, 
                                     genome <span class="op">=</span> <span class="va">Hsapiens</span>,
                                     genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000070371"</span>,
                                               <span class="st">"ENSG00000093010"</span><span class="op">)</span>, 
                                     nCores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Predicting coverage for 23 transcripts...</span>
<span class="co">#&gt; ENST00000361682</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000403184</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000403710</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000407537</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000467943</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000412786</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000207636</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000406520</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000449653</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000493893</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000428707</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000621271</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000427926</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000412649</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000622493</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000617926</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000615606</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000617103</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000538828</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000611723</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000458188</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000540896</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>
<span class="co">#&gt; ENST00000449918</span>
<span class="co">#&gt; Warning in .local(x, row.names, optional, ...): 'optional' argument was ignored</span>
<span class="co">#&gt; Predicting junction coverages...</span>

<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">predCovProfiles</span><span class="op">)</span>
<span class="co">#&gt; [1] 23</span></code></pre></div>
<p>The output of <code>predictTxCoverage</code> is a list with one element per reference transcript annotated to any of the selected genes. Each element, in turn, is a list with five components:</p>
<ul>
<li>
<code>predCov</code>: an <code>Rle</code> with the predicted coverage at each position in the transcript.</li>
<li>
<code>strand</code>: the strand of the transcript.</li>
<li>
<code>junctionCov</code>: a <code>GRanges</code> object with the location and predicted coverage of each junction in the transcript.</li>
<li>
<code>aveFragLength</code>: the estimated average fragment length.</li>
<li>
<code>note</code>: either ‘covOK’, ‘covError’ or ‘covNA’. If ‘covOK’, <em><a href="https://bioconductor.org/packages/3.15/alpine" class="external-link">alpine</a></em> was able to predict a coverage profile. If ‘covError’ or ‘covNA’, the coverage profile could not be predicted, most likely because the transcript is shorter than the fragment length or because no reads in the BAM file overlapped the transcript. In both these cases, we impose a uniform coverage profile for the transcript.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predCovProfiles</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>
<span class="co">#&gt; $ENST00000403184</span>
<span class="co">#&gt; $ENST00000403184$predCov</span>
<span class="co">#&gt; numeric-Rle of length 2216 with 2216 runs</span>
<span class="co">#&gt;   Lengths:           1           1           1 ...           1           1</span>
<span class="co">#&gt;   Values :   0.0000000   0.0614766   0.7640371 ... 0.002547412 0.000580614</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ENST00000403184$strand</span>
<span class="co">#&gt; [1] "+"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ENST00000403184$junctionCov</span>
<span class="co">#&gt; GRanges object with 4 ranges and 1 metadata column:</span>
<span class="co">#&gt;       seqnames            ranges strand |  pred.cov</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]       22 19941898-19961198      + |  2.794797</span>
<span class="co">#&gt;   [2]       22 19961290-19962526      + |  0.435794</span>
<span class="co">#&gt;   [3]       22 19962816-19963565      + |  0.814605</span>
<span class="co">#&gt;   [4]       22 19963760-19964167      + |  0.426161</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ENST00000403184$aveFragLength</span>
<span class="co">#&gt; [1] 187.9594</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ENST00000403184$note</span>
<span class="co">#&gt; [1] "covOK"</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="scale-transcript-coverage-profiles">Scale transcript coverage profiles<a class="anchor" aria-label="anchor" href="#scale-transcript-coverage-profiles"></a>
</h1>
<p>Based on the predicted transcript coverage profiles obtained above, and a set of estimated transcript abundances, we next determine the number of reads predicted to map across each annotated splice junction.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Read transcript abundance estimates</span>
<span class="va">txQuants</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/quant.sub.rds"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Scale predicted coverages based on the estimated abundances</span>
<span class="va">txsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleTxCoverages.html">scaleTxCoverages</a></span><span class="op">(</span>txCoverageProfiles <span class="op">=</span> <span class="va">predCovProfiles</span>, 
                         txQuants <span class="op">=</span> <span class="va">txQuants</span>, tx2gene <span class="op">=</span> <span class="va">tx2gene</span>,
                         strandSpecific <span class="op">=</span> <span class="cn">TRUE</span>, methodName <span class="op">=</span> <span class="st">"Salmon"</span>, 
                         verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><code>scaleTxCoverages</code> returns a list with two elements:</p>
<ul>
<li>
<code>junctionPredCovs</code>: A <code>data.frame</code> with the predicted coverage for each junction, scaled by the estimated transcript abundances and summarized across all transcripts.</li>
<li>
<code>txQuants</code>: A <code>data.frame</code> with estimated transcript abundances, including information about whether or not the coverage profile could be predicted (‘covOK’) or if a uniform coverage was imposed (‘covError’ or ‘covNA’).</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">txsc</span><span class="op">)</span>
<span class="co">#&gt; [1] "junctionPredCovs" "txQuants"</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">txsc</span><span class="op">$</span><span class="va">junctionPredCovs</span><span class="op">)</span>
<span class="co">#&gt;   seqnames    start      end width strand   predCov</span>
<span class="co">#&gt; 1       22 19941898 19961198 19301      +  75.58930</span>
<span class="co">#&gt; 2       22 19961290 19962526  1237      +  10.66403</span>
<span class="co">#&gt; 3       22 19962816 19963565   750      + 339.82310</span>
<span class="co">#&gt; 4       22 19963760 19964167   408      + 181.68201</span>
<span class="co">#&gt; 5       22 19964300 19968535  4236      + 159.39402</span>
<span class="co">#&gt; 6       22 19961290 19962232   943      +  50.41439</span>
<span class="co">#&gt;                                                                                                                                        transcript</span>
<span class="co">#&gt; 1                                                                 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000467943</span>
<span class="co">#&gt; 2                                                 ENST00000361682,ENST00000403184,ENST00000467943,ENST00000412786,ENST00000207636,ENST00000406520</span>
<span class="co">#&gt; 3                 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000207636,ENST00000406520,ENST00000449653</span>
<span class="co">#&gt; 4 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000406520,ENST00000449653,ENST00000493893,ENST00000428707</span>
<span class="co">#&gt; 5                                 ENST00000361682,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000207636,ENST00000406520,ENST00000449653</span>
<span class="co">#&gt; 6                                                                                                                                 ENST00000403710</span>
<span class="co">#&gt;              gene method</span>
<span class="co">#&gt; 1 ENSG00000093010 Salmon</span>
<span class="co">#&gt; 2 ENSG00000093010 Salmon</span>
<span class="co">#&gt; 3 ENSG00000093010 Salmon</span>
<span class="co">#&gt; 4 ENSG00000093010 Salmon</span>
<span class="co">#&gt; 5 ENSG00000093010 Salmon</span>
<span class="co">#&gt; 6 ENSG00000093010 Salmon</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">txsc</span><span class="op">$</span><span class="va">txQuants</span><span class="op">)</span>
<span class="co">#&gt;        transcript            gene   count      TPM method covNote</span>
<span class="co">#&gt; 1 ENST00000621271 ENSG00000070371 263.734 0.748398 Salmon   covOK</span>
<span class="co">#&gt; 2 ENST00000427926 ENSG00000070371 170.231 0.462868 Salmon   covOK</span>
<span class="co">#&gt; 3 ENST00000412649 ENSG00000070371  59.429 0.259538 Salmon   covOK</span>
<span class="co">#&gt; 4 ENST00000622493 ENSG00000070371  39.974 0.415456 Salmon   covOK</span>
<span class="co">#&gt; 5 ENST00000617926 ENSG00000070371   0.000 0.000000 Salmon   covOK</span>
<span class="co">#&gt; 6 ENST00000615606 ENSG00000070371   0.000 0.000000 Salmon   covOK</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="combine-observed-and-predicted-junction-coverages">Combine observed and predicted junction coverages<a class="anchor" aria-label="anchor" href="#combine-observed-and-predicted-junction-coverages"></a>
</h1>
<p>In order to calculate JCC scores, we need to match the predicted junction coverages obtained above with the observed coverages obtained by aligning the reads to the genome. Assuming the reads were aligned to the genome with STAR, we have a text file (whose name ends with <code>SJ.out.tab</code>) containing the number of reads spanning each annotated junction. The code below shows how to read this file, add column names and combine these observed junction coverages with the predicted ones obtained above. If the reads were aligned to the genome with another aligner than STAR, the observed junction coverages may need to be determined separately and formatted in the appropriate way.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Read the SJ.out.tab file from STAR, add column names and encode the strand</span>
<span class="co">## information as +/-</span>
<span class="va">jcov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sub.SJ.out.tab"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span>,
                   header <span class="op">=</span> <span class="cn">FALSE</span>, as.is <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqnames"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"strand"</span>, <span class="st">"motif"</span>, <span class="st">"annot"</span>,
             <span class="st">"uniqreads"</span>, <span class="st">"mmreads"</span>, <span class="st">"maxoverhang"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span><span class="op">(</span><span class="va">strand</span>, <span class="va">strand</span> <span class="op">==</span> <span class="fl">1</span>, <span class="st">"+"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span><span class="op">(</span><span class="va">strand</span>, <span class="va">strand</span> <span class="op">==</span> <span class="fl">2</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">seqnames</span>, <span class="va">start</span>, <span class="va">end</span>, <span class="va">strand</span>, <span class="va">uniqreads</span>, <span class="va">mmreads</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seqnames</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">jcov</span><span class="op">)</span>
<span class="co">#&gt;   seqnames    start      end strand uniqreads mmreads</span>
<span class="co">#&gt; 1       22 10848028 10870192      +         0       1</span>
<span class="co">#&gt; 2       22 10941781 10944966      -         3       1</span>
<span class="co">#&gt; 3       22 10947419 10949211      -         4       0</span>
<span class="co">#&gt; 4       22 10949270 10950048      -         1       0</span>
<span class="co">#&gt; 5       22 10950136 10959066      -         3       0</span>
<span class="co">#&gt; 6       22 10950175 10959066      -         2      11</span>

<span class="co">## Combine the observed and predicted junction coverages</span>
<span class="va">combCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineCoverages.html">combineCoverages</a></span><span class="op">(</span>junctionCounts <span class="op">=</span> <span class="va">jcov</span>, 
                            junctionPredCovs <span class="op">=</span> <span class="va">txsc</span><span class="op">$</span><span class="va">junctionPredCovs</span>,
                            txQuants <span class="op">=</span> <span class="va">txsc</span><span class="op">$</span><span class="va">txQuants</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = c("seqnames", "start", "end", "gene")</span></code></pre></div>
<p>The output of <code>combineCoverages</code> is a list with three elements:</p>
<ul>
<li>
<code>junctionCovs</code>: A <code>data.frame</code> with the observed and predicted coverages for each junction, aggregated across all transcripts. The <code>predCov</code> column contains the predicted number of spanning reads, and the <code>uniqreads</code> column will be used to represent the observed number of junction-spanning reads.</li>
<li>
<code>txQuants</code>: A <code>data.frame</code> with estimated transcript abundances, including information about whether or not the coverage profile could be predicted (‘covOK’) or if a uniform coverage was imposed (‘covError’ or ‘covNA’).</li>
<li>
<code>geneQuants</code>: A <code>data.frame</code> with estimated gene abundances as well as the total number uniquely mapping and multi-mapping reads spanning any of the junctions in the gene.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">combCov</span><span class="op">)</span>
<span class="co">#&gt; [1] "junctionCovs" "txQuants"     "geneQuants"</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combCov</span><span class="op">$</span><span class="va">junctionCovs</span><span class="op">)</span>
<span class="co">#&gt;   junctionid seqnames    start      end width strand   predCov</span>
<span class="co">#&gt; 1         J1       22 19941898 19961198 19301      +  75.58930</span>
<span class="co">#&gt; 2         J4       22 19961290 19962526  1237      +  10.66403</span>
<span class="co">#&gt; 3         J8       22 19962816 19963565   750      + 339.82310</span>
<span class="co">#&gt; 4        J10       22 19963760 19964167   408      + 181.68201</span>
<span class="co">#&gt; 5        J12       22 19964300 19968535  4236      + 159.39402</span>
<span class="co">#&gt; 6         J5       22 19961290 19962232   943      +  50.41439</span>
<span class="co">#&gt;                                                                                                                                        transcript</span>
<span class="co">#&gt; 1                                                                 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000467943</span>
<span class="co">#&gt; 2                                                 ENST00000361682,ENST00000403184,ENST00000467943,ENST00000412786,ENST00000207636,ENST00000406520</span>
<span class="co">#&gt; 3                 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000207636,ENST00000406520,ENST00000449653</span>
<span class="co">#&gt; 4 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000406520,ENST00000449653,ENST00000493893,ENST00000428707</span>
<span class="co">#&gt; 5                                 ENST00000361682,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000207636,ENST00000406520,ENST00000449653</span>
<span class="co">#&gt; 6                                                                                                                                 ENST00000403710</span>
<span class="co">#&gt;              gene method uniqreads mmreads fracunique</span>
<span class="co">#&gt; 1 ENSG00000093010 Salmon        70       2  0.9722222</span>
<span class="co">#&gt; 2 ENSG00000093010 Salmon        74       0  1.0000000</span>
<span class="co">#&gt; 3 ENSG00000093010 Salmon       160       0  1.0000000</span>
<span class="co">#&gt; 4 ENSG00000093010 Salmon       365       0  1.0000000</span>
<span class="co">#&gt; 5 ENSG00000093010 Salmon       402       0  1.0000000</span>
<span class="co">#&gt; 6 ENSG00000093010 Salmon         0       0  1.0000000</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combCov</span><span class="op">$</span><span class="va">txQuants</span><span class="op">)</span>
<span class="co">#&gt;        transcript            gene   count      TPM method covNote</span>
<span class="co">#&gt; 1 ENST00000621271 ENSG00000070371 263.734 0.748398 Salmon   covOK</span>
<span class="co">#&gt; 2 ENST00000427926 ENSG00000070371 170.231 0.462868 Salmon   covOK</span>
<span class="co">#&gt; 3 ENST00000412649 ENSG00000070371  59.429 0.259538 Salmon   covOK</span>
<span class="co">#&gt; 4 ENST00000622493 ENSG00000070371  39.974 0.415456 Salmon   covOK</span>
<span class="co">#&gt; 5 ENST00000617926 ENSG00000070371   0.000 0.000000 Salmon   covOK</span>
<span class="co">#&gt; 6 ENST00000615606 ENSG00000070371   0.000 0.000000 Salmon   covOK</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combCov</span><span class="op">$</span><span class="va">geneQuants</span><span class="op">)</span>
<span class="co">#&gt;              gene method    count       TPM nbr_expressed_transcripts</span>
<span class="co">#&gt; 1 ENSG00000070371 Salmon  547.001  2.103382                         1</span>
<span class="co">#&gt; 2 ENSG00000093010 Salmon 1697.000 45.966095                         1</span>
<span class="co">#&gt;   nbr_expressed_transcripts_5p covOKfraction uniqjuncreads mmjuncreads</span>
<span class="co">#&gt; 1                            5             1           469           0</span>
<span class="co">#&gt; 2                            3             1          1073           4</span>
<span class="co">#&gt;   uniqjuncfraction</span>
<span class="co">#&gt; 1         1.000000</span>
<span class="co">#&gt; 2         0.996286</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="calculate-jcc-scores">Calculate JCC scores<a class="anchor" aria-label="anchor" href="#calculate-jcc-scores"></a>
</h1>
<p>Finally, we calculate scaled predicted junction coverages and summarize the differences between the observed and predicted junction coverages for each gene by the JCC (junction coverage compatibility) scores.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Calculate a JCC score for each gene</span>
<span class="va">jcc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateJCCScores.html">calculateJCCScores</a></span><span class="op">(</span>junctionCovs <span class="op">=</span> <span class="va">combCov</span><span class="op">$</span><span class="va">junctionCovs</span>, 
                          geneQuants <span class="op">=</span> <span class="va">combCov</span><span class="op">$</span><span class="va">geneQuants</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">jcc</span><span class="op">$</span><span class="va">junctionCovs</span><span class="op">)</span>
<span class="co">#&gt;   junctionid seqnames    start      end width strand   predCov</span>
<span class="co">#&gt; 1         J1       22 19941898 19961198 19301      +  75.58930</span>
<span class="co">#&gt; 2         J4       22 19961290 19962526  1237      +  10.66403</span>
<span class="co">#&gt; 3         J8       22 19962816 19963565   750      + 339.82310</span>
<span class="co">#&gt; 4        J10       22 19963760 19964167   408      + 181.68201</span>
<span class="co">#&gt; 5        J12       22 19964300 19968535  4236      + 159.39402</span>
<span class="co">#&gt; 6         J5       22 19961290 19962232   943      +  50.41439</span>
<span class="co">#&gt;                                                                                                                                        transcript</span>
<span class="co">#&gt; 1                                                                 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000467943</span>
<span class="co">#&gt; 2                                                 ENST00000361682,ENST00000403184,ENST00000467943,ENST00000412786,ENST00000207636,ENST00000406520</span>
<span class="co">#&gt; 3                 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000207636,ENST00000406520,ENST00000449653</span>
<span class="co">#&gt; 4 ENST00000361682,ENST00000403184,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000406520,ENST00000449653,ENST00000493893,ENST00000428707</span>
<span class="co">#&gt; 5                                 ENST00000361682,ENST00000403710,ENST00000407537,ENST00000412786,ENST00000207636,ENST00000406520,ENST00000449653</span>
<span class="co">#&gt; 6                                                                                                                                 ENST00000403710</span>
<span class="co">#&gt;              gene method uniqreads mmreads fracunique jccscore scaledCov</span>
<span class="co">#&gt; 1 ENSG00000093010 Salmon        70       2  0.9722222     0.73  96.15727</span>
<span class="co">#&gt; 2 ENSG00000093010 Salmon        74       0  1.0000000     0.73  13.56573</span>
<span class="co">#&gt; 3 ENSG00000093010 Salmon       160       0  1.0000000     0.73 432.28947</span>
<span class="co">#&gt; 4 ENSG00000093010 Salmon       365       0  1.0000000     0.73 231.11796</span>
<span class="co">#&gt; 5 ENSG00000093010 Salmon       402       0  1.0000000     0.73 202.76537</span>
<span class="co">#&gt; 6 ENSG00000093010 Salmon         0       0  1.0000000     0.73  64.13222</span>
<span class="co">#&gt;     methodscore</span>
<span class="co">#&gt; 1 Salmon (0.73)</span>
<span class="co">#&gt; 2 Salmon (0.73)</span>
<span class="co">#&gt; 3 Salmon (0.73)</span>
<span class="co">#&gt; 4 Salmon (0.73)</span>
<span class="co">#&gt; 5 Salmon (0.73)</span>
<span class="co">#&gt; 6 Salmon (0.73)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">jcc</span><span class="op">$</span><span class="va">geneScores</span><span class="op">)</span>
<span class="co">#&gt;              gene method    count       TPM nbr_expressed_transcripts</span>
<span class="co">#&gt; 1 ENSG00000070371 Salmon  547.001  2.103382                         1</span>
<span class="co">#&gt; 2 ENSG00000093010 Salmon 1697.000 45.966095                         1</span>
<span class="co">#&gt;   nbr_expressed_transcripts_5p covOKfraction uniqjuncreads mmjuncreads</span>
<span class="co">#&gt; 1                            5             1           469           0</span>
<span class="co">#&gt; 2                            3             1          1073           4</span>
<span class="co">#&gt;   uniqjuncfraction jccscore</span>
<span class="co">#&gt; 1         1.000000     0.52</span>
<span class="co">#&gt; 2         0.996286     0.73</span></code></pre></div>
<p><code>calculateJCCScores</code> returns a list with two elements:</p>
<ul>
<li>
<code>junctionCovs</code>: A <code>data.frame</code> with observed and predicted junction coverages, as well as scaled predicted coverages (<code>scaledCov</code> column) and JCC scores for the corresponding gene (<code>jccscore</code>).</li>
<li>
<code>geneScores</code>: A <code>data.frame</code> with gene-level abundances, extended with the calculated JCC scores (<code>jccscore</code> column).</li>
</ul>
</div>
<div class="section level1">
<h1 id="plot-observed-and-predicted-junction-coverages">Plot observed and predicted junction coverages<a class="anchor" aria-label="anchor" href="#plot-observed-and-predicted-junction-coverages"></a>
</h1>
<p>To investigate the agreement between the observed and predicted junction coverages for a given gene, we can plot the <code>uniqreads</code> and <code>scaledCov</code> columns from the junction coverage data frame generated in the previous step:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">selGene</span> <span class="op">&lt;-</span> <span class="st">"ENSG00000070371"</span>
<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">jcc</span><span class="op">$</span><span class="va">junctionCovs</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
                  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op">==</span> <span class="va">selGene</span><span class="op">)</span>, 
                <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uniqreads</span>, y <span class="op">=</span> <span class="va">scaledCov</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept <span class="op">=</span> <span class="fl">0</span>, slope <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Number of uniquely mapped reads spanning junction"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Scaled predicted junction coverage"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">jcc</span><span class="op">$</span><span class="va">junctionCovs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
                            <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op">==</span> <span class="va">selGene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
                            <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">methodscore</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="JCCscore_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>The package also contains a function to generate a multi-panel plot, showing (i) the observed genome coverage profile in the region of the gene, (ii) the estimated transcript abundances, and (iii) the association between the observed and predicted junction coverages.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/reads.chr22.bw"</span>, package <span class="op">=</span> <span class="st">"jcc"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotGeneSummary.html">plotGeneSummary</a></span><span class="op">(</span>gtf <span class="op">=</span> <span class="va">gtf</span>, junctionCovs <span class="op">=</span> <span class="va">jcc</span><span class="op">$</span><span class="va">junctionCovs</span>, 
                useGene <span class="op">=</span> <span class="st">"ENSG00000070371"</span>, bwFile <span class="op">=</span> <span class="va">bwFile</span>,
                txQuants <span class="op">=</span> <span class="va">combCov</span><span class="op">$</span><span class="va">txQuants</span>, txCol <span class="op">=</span> <span class="st">"transcript_id"</span>,
                geneCol <span class="op">=</span> <span class="st">"gene_id"</span>, exonCol <span class="op">=</span> <span class="st">"exon_id"</span><span class="op">)</span>
<span class="co">#&gt; Warning: Package `magick` is required to draw images. Image not drawn.</span></code></pre></div>
<p><img src="JCCscore_files/figure-html/unnamed-chunk-13-1.png" width="1152"></p>
</div>
<div class="section level1">
<h1 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2021-10-29 r81110)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] BSgenome.Hsapiens.NCBI.GRCh38_1.3.1000</span>
<span class="co">#&gt;  [2] BSgenome_1.63.1                       </span>
<span class="co">#&gt;  [3] rtracklayer_1.55.0                    </span>
<span class="co">#&gt;  [4] Biostrings_2.63.0                     </span>
<span class="co">#&gt;  [5] XVector_0.35.0                        </span>
<span class="co">#&gt;  [6] GenomicRanges_1.46.0                  </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.31.1                   </span>
<span class="co">#&gt;  [8] IRanges_2.29.0                        </span>
<span class="co">#&gt;  [9] S4Vectors_0.33.0                      </span>
<span class="co">#&gt; [10] BiocGenerics_0.41.1                   </span>
<span class="co">#&gt; [11] dplyr_1.0.7                           </span>
<span class="co">#&gt; [12] jcc_0.1.0                             </span>
<span class="co">#&gt; [13] BiocStyle_2.23.0                      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] backports_1.3.0             Hmisc_4.6-0                </span>
<span class="co">#&gt;   [3] BiocFileCache_2.3.0         systemfonts_1.0.3          </span>
<span class="co">#&gt;   [5] lazyeval_0.2.2              splines_4.2.0              </span>
<span class="co">#&gt;   [7] BiocParallel_1.29.0         ggplot2_3.3.5              </span>
<span class="co">#&gt;   [9] digest_0.6.28               ensembldb_2.19.0           </span>
<span class="co">#&gt;  [11] htmltools_0.5.2             fansi_0.5.0                </span>
<span class="co">#&gt;  [13] magrittr_2.0.1              checkmate_2.0.0            </span>
<span class="co">#&gt;  [15] memoise_2.0.0               cluster_2.1.2              </span>
<span class="co">#&gt;  [17] matrixStats_0.61.0          pkgdown_1.9000.9000.9000   </span>
<span class="co">#&gt;  [19] prettyunits_1.1.1           jpeg_0.1-9                 </span>
<span class="co">#&gt;  [21] colorspace_2.0-2            blob_1.2.2                 </span>
<span class="co">#&gt;  [23] rappdirs_0.3.3              textshaping_0.3.6          </span>
<span class="co">#&gt;  [25] xfun_0.27                   crayon_1.4.2               </span>
<span class="co">#&gt;  [27] RCurl_1.98-1.5              jsonlite_1.7.2             </span>
<span class="co">#&gt;  [29] graph_1.73.0                scatterpie_0.1.7           </span>
<span class="co">#&gt;  [31] survival_3.2-13             VariantAnnotation_1.41.0   </span>
<span class="co">#&gt;  [33] glue_1.4.2                  polyclip_1.10-0            </span>
<span class="co">#&gt;  [35] gtable_0.3.0                zlibbioc_1.41.0            </span>
<span class="co">#&gt;  [37] alpine_1.21.0               DelayedArray_0.21.0        </span>
<span class="co">#&gt;  [39] scales_1.1.1                DBI_1.1.1                  </span>
<span class="co">#&gt;  [41] Rcpp_1.0.7                  progress_1.2.2             </span>
<span class="co">#&gt;  [43] htmlTable_2.3.0             foreign_0.8-81             </span>
<span class="co">#&gt;  [45] bit_4.0.4                   Formula_1.2-4              </span>
<span class="co">#&gt;  [47] htmlwidgets_1.5.4           httr_1.4.2                 </span>
<span class="co">#&gt;  [49] RColorBrewer_1.1-2          speedglm_0.3-3             </span>
<span class="co">#&gt;  [51] ellipsis_0.3.2              pkgconfig_2.0.3            </span>
<span class="co">#&gt;  [53] XML_3.99-0.8                farver_2.1.0               </span>
<span class="co">#&gt;  [55] Gviz_1.39.0                 nnet_7.3-16                </span>
<span class="co">#&gt;  [57] sass_0.4.0                  dbplyr_2.1.1               </span>
<span class="co">#&gt;  [59] utf8_1.2.2                  labeling_0.4.2             </span>
<span class="co">#&gt;  [61] tidyselect_1.1.1            rlang_0.99.0.9000          </span>
<span class="co">#&gt;  [63] AnnotationDbi_1.57.0        munsell_0.5.0              </span>
<span class="co">#&gt;  [65] tools_4.2.0                 cachem_1.0.6               </span>
<span class="co">#&gt;  [67] cli_3.1.0                   generics_0.1.1             </span>
<span class="co">#&gt;  [69] RSQLite_2.2.8               evaluate_0.14              </span>
<span class="co">#&gt;  [71] stringr_1.4.0               fastmap_1.1.0              </span>
<span class="co">#&gt;  [73] yaml_2.2.1                  ragg_1.1.3                 </span>
<span class="co">#&gt;  [75] knitr_1.36                  bit64_4.0.5                </span>
<span class="co">#&gt;  [77] fs_1.5.0                    purrr_0.3.4                </span>
<span class="co">#&gt;  [79] KEGGREST_1.35.0             AnnotationFilter_1.19.0    </span>
<span class="co">#&gt;  [81] RBGL_1.71.0                 xml2_1.3.2                 </span>
<span class="co">#&gt;  [83] biomaRt_2.51.0              compiler_4.2.0             </span>
<span class="co">#&gt;  [85] rstudioapi_0.13             filelock_1.0.2             </span>
<span class="co">#&gt;  [87] curl_4.3.2                  png_0.1-7                  </span>
<span class="co">#&gt;  [89] tibble_3.1.5                tweenr_1.0.2               </span>
<span class="co">#&gt;  [91] bslib_0.3.1                 stringi_1.7.5              </span>
<span class="co">#&gt;  [93] highr_0.9                   GenomicFeatures_1.47.1     </span>
<span class="co">#&gt;  [95] desc_1.4.0                  lattice_0.20-45            </span>
<span class="co">#&gt;  [97] ProtGenerics_1.27.0         Matrix_1.3-4               </span>
<span class="co">#&gt;  [99] vctrs_0.3.8                 pillar_1.6.4               </span>
<span class="co">#&gt; [101] lifecycle_1.0.1             BiocManager_1.30.16        </span>
<span class="co">#&gt; [103] jquerylib_0.1.4             data.table_1.14.2          </span>
<span class="co">#&gt; [105] cowplot_1.1.1               bitops_1.0-7               </span>
<span class="co">#&gt; [107] R6_2.5.1                    BiocIO_1.5.0               </span>
<span class="co">#&gt; [109] latticeExtra_0.6-29         bookdown_0.24              </span>
<span class="co">#&gt; [111] gridExtra_2.3               dichromat_2.0-0            </span>
<span class="co">#&gt; [113] MASS_7.3-54                 assertthat_0.2.1           </span>
<span class="co">#&gt; [115] SummarizedExperiment_1.25.0 rprojroot_2.0.2            </span>
<span class="co">#&gt; [117] rjson_0.2.20                GenomicAlignments_1.31.0   </span>
<span class="co">#&gt; [119] Rsamtools_2.11.0            GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt; [121] parallel_4.2.0              hms_1.1.1                  </span>
<span class="co">#&gt; [123] grid_4.2.0                  rpart_4.1-15               </span>
<span class="co">#&gt; [125] ggfun_0.0.4                 tidyr_1.1.4                </span>
<span class="co">#&gt; [127] rmarkdown_2.11              MatrixGenerics_1.7.0       </span>
<span class="co">#&gt; [129] biovizBase_1.43.0           ggforce_0.3.3              </span>
<span class="co">#&gt; [131] Biobase_2.55.0              base64enc_0.1-3            </span>
<span class="co">#&gt; [133] restfulr_0.0.13</span></code></pre></div>
</div>
<div class="section level1 unnumbered">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references">
<div id="ref-dobin2013star">
<p>Dobin, Alexander, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. 2013. “STAR: Ultrafast Universal RNA-seq Aligner.” <em>Bioinformatics</em> 29 (1): 15–21.</p>
</div>
<div id="ref-love2016alpine">
<p>Love, Michael I, John B Hogenesch, and Rafael A Irizarry. 2016. “Modeling of RNA-seq Fragment Sequence Bias Reduces Systematic Errors in Transcript Abundance Estimation.” <em>Nat. Biotechnol.</em> 34 (12): 1287–91.</p>
</div>
<div id="ref-soneson2018jcc">
<p>Soneson, Charlotte, Michael I Love, Rob Patro, Shobbir Hussain, Dheeraj Malhotra, and Mark D Robinson. 2018. “A Junction Coverage Compatibility Score to Quantify the Reliability of Transcript Abundance Estimates and Annotation Catalogs.” <em>bioRxiv 10.1101/378539</em>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Charlotte Soneson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
